Description: The QL-specific command (ESC i A "Enable Cutter") needs to
 be sent to the printer so that cutting happens, but it only needs to be
 sent to QL-series device.
Bugs-Ubuntu: LP: #635249.
Bugs-Ubuntu: https://bugs.launchpad.net/bugs/1287971
Bugs-Debian: https://bugs.debian.org/740789
Author: Till Kamppeter <till.kamppeter@gmail.com>
Author: Philip Pemberton <philpem@philpem.me.uk>
Last-Update: 2014-03-05
--- a/rastertoptch.c
+++ b/rastertoptch.c
@@ -685,7 +685,8 @@
 inline void
 emit_feed_cut_mirror (bool do_feed, unsigned feed,
                       bool do_cut,
-                      bool do_mirror) {
+                      bool do_mirror,
+                      xfer_t xfer_mode) {
   /* Determine feed nibble */
   unsigned feed_nibble;
   if (do_feed) {
@@ -698,6 +699,14 @@
   /* Determine mirror print bit*/
   unsigned char mirror_bit = do_mirror ? 0x80 : 0x00;
   /* Combine & emit printer command code */
+  if (xfer_mode == ULP) {
+    /* ESC i A is Enable Cutter -- used for QL-560 only, according to
+     * <http://www.undocprint.org/formats/page_description_languages/brother_p-touch>
+     * The QL-560 (actually the whole QL series) uses ULP mode, so we check for that.
+     * The PT2450DX uses RLE and throw an INTERFACE ERROR if it sees this command.
+     */
+    putchar (ESC); putchar ('i'); putchar ('A'); putchar ((char) (do_cut ? 0x01 : 0x00));
+  }
   putchar (ESC); putchar ('i'); putchar ('M');
   putchar ((char) (feed & 0x1f) | auto_cut_bit | mirror_bit);
 }
@@ -806,7 +815,8 @@
     /* We only know how to feed after each page */
     emit_feed_cut_mirror (perform_feed == CUPS_ADVANCE_PAGE, feed,
                           cut_media == CUPS_CUT_PAGE,
-                          mirror == CUPS_TRUE);
+                          mirror == CUPS_TRUE,
+                          job_options->pixel_xfer == ULP);
   /* Set media and quality if label preamble is requested */
   unsigned page_size_y = new_page_options->page_size [1];
   unsigned image_height_px = lrint (page_size_y * vres / 72.0);
@@ -1526,7 +1536,8 @@
            perform_feed == CUPS_ADVANCE_JOB,
            new_page_options->feed,
            cut_media == CUPS_CUT_PAGE || cut_media == CUPS_CUT_JOB,
-           new_page_options->mirror == CUPS_TRUE);
+           new_page_options->mirror == CUPS_TRUE,
+           job_options->pixel_xfer);
         /* Emit eject marker */
         putchar (PTC_EJECT);
       }
